core/domain/exp_jobs_one_off.py

```diff
@@ -164,120 +248,163 @@ class MathExpressionValidationOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+                            elif is_valid_math_equation(rule_input):
+                                type_of_input = _TYPE_VALID_MATH_EQUATION
+
+                            types_of_inputs.add(type_of_input)
+
+                            if type_of_input != _TYPE_INVALID_EXPRESSION:
+                                rule_spec.inputs['x'] = rule_input
+                                if type_of_input == _TYPE_VALID_MATH_EQUATION:
+                                    rule_spec.inputs['y'] = 'both'
+                                rule_spec.rule_type = 'MatchesExactlyWith'
 
                             output_values = '%s %s: %s' % (
                                 item.id, state_name, rule_input)
 
-                            yield (validity, output_values.encode('utf-8'))
+                            yield (type_of_input, output_values.encode('utf-8'))
+
+                        new_answer_groups.append(new_answer_group.to_dict())
+
+                    if _TYPE_INVALID_EXPRESSION in types_of_inputs:
+                        yield ('ERROR', (
+                            'There are some invalid inputs that need to be '
+                            'resolved before running the upgrade job.'))
+                    elif len(types_of_inputs) > 1:
+                        yield ('ERROR', (
+                            'The exploration with ID: %s and state name: %s '
+                            'contains inputs that correspond to multiple '
+                            'different types: %s. Please resolve this before '
+                            'running the upgrade job.' % (
+                                item.id, state_name,
+                                ', '.join(list(types_of_inputs)))))
+                    else:
+                        exp_services.update_exploration(
+                            feconf.MIGRATION_BOT_USER_ID, item.id, [
+                                exp_domain.ExplorationChange({
+                                    'cmd': 'edit_state_property',
+                                    'state_name': state_name,
+                                    'property_name': 'widget_id',
+                                    'new_value': list(types_of_inputs)[0]
+                                }), exp_domain.ExplorationChange({
+                                    'cmd': 'edit_state_property',
+                                    'state_name': state_name,
+                                    'property_name': 'answer_groups',
+                                    'new_value': new_answer_groups
+                                })], 'Upgraded interaction to %s.' % (
+                                    list(types_of_inputs)[0]))
 
     @staticmethod
     def reduce(key, values):
-        valid_inputs_limit = (
-            MathExpressionValidationOneOffJob.VALID_MATH_INPUTS_YIELD_LIMIT)
-        if key.startswith('Valid'):
-            yield (key, values[:valid_inputs_limit])
+        if key.endswith('Input'):
```

Comment:
Be more specific here. If you're trying to say "if key is one of _TYPE_..." then say that. Otherwise it's hard for the reader of this to know what your intention is, and "Input" itself doesn't really convey much meaning compared to "Valid"/"Invalid".

Reply:
Done.